<!DOCTYPE html>
<html lang="es">
<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Conceptos Básicos de Manejo de Errores</title>

    <link rel="alternate" href="https://campusempresa.com/mod/go/04-01-error-handling-basics" hreflang="es" />
	<link rel="alternate" href="https://campusempresa.cat/mod/go/04-01-error-handling-basics" hreflang="ca" />
	<link rel="alternate" href="https://enterprisecampus.net/mod/go/04-01-error-handling-basics" hreflang="en" />
    
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap-icons/font/bootstrap-icons.css">
	<link href="/css/site.css" rel="stylesheet">
	
    <script src="https://code.jquery.com/jquery-3.7.1.min.js"></script>
  	<script type="text/javascript" src="/js/math_init.js"></script>
  	<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/startup.js"></script>
  	<script type="text/javascript" src="/js/cookie.js"></script>
  	<script type="text/javascript" src="/js/main.js"></script>
</head>

<body >
    <div id="content">
		<div id="header" class="container-xxl">
	<div class="row">
		<div class="col-12 col-md-6 p-2 p-md-0">
			<h1 class="m-0 p-0">
				<a href="/"><img src="/img/logo_header.png"></a>
			</h1>
		</div>
		<div class="col-12 col-md-6 p-2 p-md-0 text-end">
				<a href="https://enterprisecampus.net/mod/go/04-01-error-handling-basics" class="px-2">EN</a></b>
	|
	<b class="px-2">ES</b>
	|
	<a href="https://campusempresa.cat/mod/go/04-01-error-handling-basics" class="px-2">CA</a>
<br>
			<cite>Construyendo la sociedad de hoy y del mañana</cite>
		</div>
	</div>
</div>
<div id="subheader" class="container-xxl">
	<div class="row">
		<div class="col-12 p-2 p-md-0 m-0 text-end">
			<a href="/objective">El Proyecto</a> | 
<a href="/about">Sobre nosotros</a> | 
<a href="/contribute">Contribuir</a> | 
<a href="/donate">Donaciones</a> | 
<a href="/licence">Licencia</a>
		</div>
	</div>
</div>

<div class="top-bar container-fluid">
	<div class="container-xxl">
		<div class="row">
			<div class="col" id="left_menu">
									<a href="./">Contenido del curso</a>
					<span class="sep">|</span>
								<a href="/all/competencias">Competencias técnicas</a>
				<a href="/all/conocimientos">Conocimientos</a>
				<a href="/all/soft_skills">Competencias sociales</a>
			</div>
		</div>
	</div>
</div>

<div class="container-xxl" id="main_content">
	<div class="row">
		<div class="col-12 col-lg-8">
			<div id="nav1" class="navigation"></div>
			<div id="inner_content">
								<div class='row navigation'>
	<div class='col-1 col-md-2'>
					<a href='03-04-pointers' title="Punteros">
				<span class="d-none d-md-inline">&#x25C4; Anterior</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-left-square-fill"></i></span>
			</a>
			</div>
	<div class='col-10 col-md-8 text-center'>
					<a href="./"><h2 style="text-decoration:underline">Conceptos Básicos de Manejo de Errores</h2></a>
			</div>
	<div class='col-1 col-md-2 text-end'>
					<a href='04-02-custom-errors' title="Errores Personalizados">
				<span class="d-none d-md-inline">Siguiente &#x25BA;</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-right-square-fill"></i></span>
			</a>
			</div>
</div>
<div class='content'><p>El manejo de errores es una parte crucial de la programación en cualquier lenguaje, y Go no es la excepción. En este módulo, aprenderás cómo manejar errores de manera efectiva en Go, lo que te permitirá escribir código más robusto y confiable.</p>
</div><h1><ol>
<li>Introducción al Manejo de Errores en Go</li>
</ol>
</h1>
<div class='content'><p>En Go, el manejo de errores se realiza de manera explícita. A diferencia de otros lenguajes que utilizan excepciones, Go utiliza valores de retorno para indicar errores. Esto hace que el manejo de errores sea más claro y predecible.</p>
</div><h2><p>Ejemplo Básico</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("cGFja2FnZSBtYWluCgppbXBvcnQgKAogICAgImVycm9ycyIKICAgICJmbXQiCikKCmZ1bmMgZGl2aWRlKGEsIGIgZmxvYXQ2NCkgKGZsb2F0NjQsIGVycm9yKSB7CiAgICBpZiBiID09IDAgewogICAgICAgIHJldHVybiAwLCBlcnJvcnMuTmV3KCJjYW5ub3QgZGl2aWRlIGJ5IHplcm8iKQogICAgfQogICAgcmV0dXJuIGEgLyBiLCBuaWwKfQoKZnVuYyBtYWluKCkgewogICAgcmVzdWx0LCBlcnIgOj0gZGl2aWRlKDQsIDApCiAgICBpZiBlcnIgIT0gbmlsIHsKICAgICAgICBmbXQuUHJpbnRsbigiRXJyb3I6IiwgZXJyKQogICAgfSBlbHNlIHsKICAgICAgICBmbXQuUHJpbnRsbigiUmVzdWx0OiIsIHJlc3VsdCkKICAgIH0KfQ=="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>package main

import (
    &quot;errors&quot;
    &quot;fmt&quot;
)

func divide(a, b float64) (float64, error) {
    if b == 0 {
        return 0, errors.New(&quot;cannot divide by zero&quot;)
    }
    return a / b, nil
}

func main() {
    result, err := divide(4, 0)
    if err != nil {
        fmt.Println(&quot;Error:&quot;, err)
    } else {
        fmt.Println(&quot;Result:&quot;, result)
    }
}</pre></div><div class='content'></div><h2><p>Explicación del Código</p>
</h2>
<div class='content'><ol>
<li><strong>Importación de Paquetes</strong>: Importamos los paquetes <code>errors</code> y <code>fmt</code>.</li>
<li><strong>Función <code>divide</code></strong>: Esta función toma dos números <code>a</code> y <code>b</code> y devuelve el resultado de la división y un error. Si <code>b</code> es 0, la función retorna un error utilizando <code>errors.New</code>.</li>
<li><strong>Manejo del Error en <code>main</code></strong>: Llamamos a la función <code>divide</code> y verificamos si hay un error. Si hay un error, lo imprimimos; de lo contrario, imprimimos el resultado.</li>
</ol>
</div><h1><ol start="2">
<li>Uso del Paquete <code>errors</code></li>
</ol>
</h1>
<div class='content'><p>El paquete <code>errors</code> proporciona una manera sencilla de crear errores. La función <code>errors.New</code> toma un mensaje de error y devuelve un valor de tipo <code>error</code>.</p>
</div><h2><p>Ejemplo</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("cGFja2FnZSBtYWluCgppbXBvcnQgKAogICAgImVycm9ycyIKICAgICJmbXQiCikKCmZ1bmMgbWFpbigpIHsKICAgIGVyciA6PSBlcnJvcnMuTmV3KCJhbiBlcnJvciBvY2N1cnJlZCIpCiAgICBmbXQuUHJpbnRsbihlcnIpCn0="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>package main

import (
    &quot;errors&quot;
    &quot;fmt&quot;
)

func main() {
    err := errors.New(&quot;an error occurred&quot;)
    fmt.Println(err)
}</pre></div><div class='content'></div><h1><ol start="3">
<li>Comprobación de Errores</li>
</ol>
</h1>
<div class='content'><p>En Go, es común comprobar los errores inmediatamente después de una operación que puede fallar. Esto se hace utilizando una declaración <code>if</code>.</p>
</div><h2><p>Ejemplo</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("cGFja2FnZSBtYWluCgppbXBvcnQgKAogICAgImZtdCIKICAgICJzdHJjb252IgopCgpmdW5jIG1haW4oKSB7CiAgICBudW0sIGVyciA6PSBzdHJjb252LkF0b2koIjEyM2EiKQogICAgaWYgZXJyICE9IG5pbCB7CiAgICAgICAgZm10LlByaW50bG4oIkVycm9yOiIsIGVycikKICAgIH0gZWxzZSB7CiAgICAgICAgZm10LlByaW50bG4oIk51bWJlcjoiLCBudW0pCiAgICB9Cn0="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>package main

import (
    &quot;fmt&quot;
    &quot;strconv&quot;
)

func main() {
    num, err := strconv.Atoi(&quot;123a&quot;)
    if err != nil {
        fmt.Println(&quot;Error:&quot;, err)
    } else {
        fmt.Println(&quot;Number:&quot;, num)
    }
}</pre></div><div class='content'></div><h2><p>Explicación del Código</p>
</h2>
<div class='content'><ol>
<li><strong>Conversión de Cadena a Entero</strong>: Intentamos convertir la cadena <code>&quot;123a&quot;</code> a un entero utilizando <code>strconv.Atoi</code>.</li>
<li><strong>Comprobación del Error</strong>: Si hay un error, lo imprimimos; de lo contrario, imprimimos el número convertido.</li>
</ol>
</div><h1><ol start="4">
<li>Errores Personalizados</li>
</ol>
</h1>
<div class='content'><p>Puedes definir tus propios tipos de error para proporcionar más contexto sobre lo que salió mal.</p>
</div><h2><p>Ejemplo</p>
</h2>
<div class='content'></div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("cGFja2FnZSBtYWluCgppbXBvcnQgKAogICAgImZtdCIKKQoKdHlwZSBNeUVycm9yIHN0cnVjdCB7CiAgICBNc2cgc3RyaW5nCn0KCmZ1bmMgKGUgKk15RXJyb3IpIEVycm9yKCkgc3RyaW5nIHsKICAgIHJldHVybiBlLk1zZwp9CgpmdW5jIGRvU29tZXRoaW5nKGZsYWcgYm9vbCkgZXJyb3IgewogICAgaWYgIWZsYWcgewogICAgICAgIHJldHVybiAmTXlFcnJvcntNc2c6ICJzb21ldGhpbmcgd2VudCB3cm9uZyJ9CiAgICB9CiAgICByZXR1cm4gbmlsCn0KCmZ1bmMgbWFpbigpIHsKICAgIGVyciA6PSBkb1NvbWV0aGluZyhmYWxzZSkKICAgIGlmIGVyciAhPSBuaWwgewogICAgICAgIGZtdC5QcmludGxuKCJFcnJvcjoiLCBlcnIpCiAgICB9Cn0="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>package main

import (
    &quot;fmt&quot;
)

type MyError struct {
    Msg string
}

func (e *MyError) Error() string {
    return e.Msg
}

func doSomething(flag bool) error {
    if !flag {
        return &amp;MyError{Msg: &quot;something went wrong&quot;}
    }
    return nil
}

func main() {
    err := doSomething(false)
    if err != nil {
        fmt.Println(&quot;Error:&quot;, err)
    }
}</pre></div><div class='content'></div><h2><p>Explicación del Código</p>
</h2>
<div class='content'><ol>
<li><strong>Definición de <code>MyError</code></strong>: Creamos un tipo <code>MyError</code> que implementa la interfaz <code>error</code>.</li>
<li><strong>Función <code>doSomething</code></strong>: Esta función retorna un error de tipo <code>MyError</code> si <code>flag</code> es <code>false</code>.</li>
<li><strong>Manejo del Error en <code>main</code></strong>: Llamamos a <code>doSomething</code> y verificamos si hay un error. Si hay un error, lo imprimimos.</li>
</ol>
</div><h1><p>Ejercicio Práctico</p>
</h1>
<div class='content'></div><h2><p>Ejercicio</p>
</h2>
<div class='content'><p>Escribe una función <code>sqrt</code> que calcule la raíz cuadrada de un número. Si el número es negativo, la función debe retornar un error.</p>
</div><div style='position:relative'><a class='copy_button' href='#' onclick='navigator.clipboard.writeText(decodeURIComponent(escape(atob("cGFja2FnZSBtYWluCgppbXBvcnQgKAogICAgImVycm9ycyIKICAgICJmbXQiCiAgICAibWF0aCIKKQoKZnVuYyBzcXJ0KHggZmxvYXQ2NCkgKGZsb2F0NjQsIGVycm9yKSB7CiAgICBpZiB4IDwgMCB7CiAgICAgICAgcmV0dXJuIDAsIGVycm9ycy5OZXcoImNhbm5vdCBjb21wdXRlIHNxdWFyZSByb290IG9mIGEgbmVnYXRpdmUgbnVtYmVyIikKICAgIH0KICAgIHJldHVybiBtYXRoLlNxcnQoeCksIG5pbAp9CgpmdW5jIG1haW4oKSB7CiAgICByZXN1bHQsIGVyciA6PSBzcXJ0KC00KQogICAgaWYgZXJyICE9IG5pbCB7CiAgICAgICAgZm10LlByaW50bG4oIkVycm9yOiIsIGVycikKICAgIH0gZWxzZSB7CiAgICAgICAgZm10LlByaW50bG4oIlJlc3VsdDoiLCByZXN1bHQpCiAgICB9Cn0="))));alert("¡Copiado!");return false;'><i class='bi bi-copy'></i></a><pre class='code'>package main

import (
    &quot;errors&quot;
    &quot;fmt&quot;
    &quot;math&quot;
)

func sqrt(x float64) (float64, error) {
    if x &lt; 0 {
        return 0, errors.New(&quot;cannot compute square root of a negative number&quot;)
    }
    return math.Sqrt(x), nil
}

func main() {
    result, err := sqrt(-4)
    if err != nil {
        fmt.Println(&quot;Error:&quot;, err)
    } else {
        fmt.Println(&quot;Result:&quot;, result)
    }
}</pre></div><div class='content'></div><h2><p>Solución</p>
</h2>
<div class='content'><ol>
<li><strong>Función <code>sqrt</code></strong>: La función toma un número <code>x</code> y retorna su raíz cuadrada y un error. Si <code>x</code> es negativo, retorna un error.</li>
<li><strong>Manejo del Error en <code>main</code></strong>: Llamamos a <code>sqrt</code> con un número negativo y verificamos si hay un error. Si hay un error, lo imprimimos; de lo contrario, imprimimos el resultado.</li>
</ol>
</div><h1><p>Conclusión</p>
</h1>
<div class='content'><p>En esta sección, hemos aprendido los conceptos básicos del manejo de errores en Go. Hemos visto cómo crear y manejar errores utilizando el paquete <code>errors</code>, cómo definir errores personalizados y cómo comprobar errores de manera efectiva. Estos conceptos son fundamentales para escribir código robusto y confiable en Go.</p>
<p>En el próximo módulo, exploraremos cómo crear errores personalizados más avanzados y cómo manejar situaciones de pánico y recuperación.</p>
</div><div class='row navigation'>
	<div class='col-1 col-md-2'>
					<a href='03-04-pointers' title="Punteros">
				<span class="d-none d-md-inline">&#x25C4; Anterior</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-left-square-fill"></i></span>
			</a>
			</div>
	<div class='col-10 col-md-8 text-center'>
			</div>
	<div class='col-1 col-md-2 text-end'>
					<a href='04-02-custom-errors' title="Errores Personalizados">
				<span class="d-none d-md-inline">Siguiente &#x25BA;</span>
				<span class="d-inline d-md-none"><i class="bi bi-caret-right-square-fill"></i></span>
			</a>
			</div>
</div>

			</div>
		</div>
		<div class="col-12 col-lg-4 publi" id="div_publi">
			<h1>Publicidad</h1>
			<p>Este espacio está destinado a publicidad.</p>
			<p>Si quieres ser patrocinador, contáctanos para incluir enlaces en esta zona: <a href='mailto:admin@campusempresa.cat'>admin@campusempresa.cat</a></p>
			<p>¡Gracias por colaborar!</p>
		</div>
	</div>
</div>

   <div class="container-xxl my-3">
	<div class="row">
		<div class="col">
			<footer>&copy; Copyright 2024. Todos los derechos reservados</footer>
		</div>
	</div>
</div>	

<div id="cookies_adv" style="display:none;">
	Usamos cookies para mejorar tu experiencia de uso y ofrecer contenidos adaptados a tus intereses.
    <a href="#" id="btn_accept_cookies" class="button">Aceptar</a>
    <a href="/cookies">Mas información</a>
</div>	

	</div>    
</body>
</html>
